test/e2e/apibinding/apibinding_test.go

```diff
@@ -51,6 +54,62 @@ import (
 	"github.com/kcp-dev/kcp/test/e2e/framework"
 )
 
+func TestAPIBindingAPIExportReferenceImmutability(t *testing.T) {
+	server := framework.SharedKcpServer(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	orgClusterName := framework.NewOrganizationFixture(t, server)
+	serviceProviderWorkspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithName("service-provider-1"))
+	consumerWorkspace := framework.NewWorkspaceFixture(t, server, orgClusterName, framework.WithName("consumer-1-bound-against-1"))
+
+	cfg := server.BaseConfig(t)
+
+	kcpClusterClient, err := clientset.NewForConfig(cfg)
+	require.NoError(t, err, "failed to construct kcp cluster client for server")
+
+	t.Logf("Create an APIExport today-cowboys in %q", serviceProviderWorkspace)
+	cowboysAPIExport := &apisv1alpha1.APIExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "today-cowboys",
+		},
+		Spec: apisv1alpha1.APIExportSpec{
+			LatestResourceSchemas: []string{"today.cowboys.wildwest.dev"},
+		},
+	}
+	cowboysAPIExport, err = kcpClusterClient.ApisV1alpha1().APIExports().Create(logicalcluster.WithCluster(ctx, serviceProviderWorkspace), cowboysAPIExport, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	t.Logf("Create an APIBinding in %q that points to the today-cowboys export from %q", consumerWorkspace, serviceProviderWorkspace)
+	apiBinding := &apisv1alpha1.APIBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cowboys",
+		},
+		Spec: apisv1alpha1.APIBindingSpec{
+			Reference: apisv1alpha1.ExportReference{
+				Workspace: &apisv1alpha1.WorkspaceExportReference{
+					Path:       serviceProviderWorkspace.String(),
+					ExportName: "today-cowboys",
+				},
+			},
+		},
+	}
+
+	_, err = kcpClusterClient.ApisV1alpha1().APIBindings().Create(logicalcluster.WithCluster(ctx, consumerWorkspace), apiBinding, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	apiBinding, err = kcpClusterClient.ApisV1alpha1().APIBindings().Get(logicalcluster.WithCluster(ctx, consumerWorkspace), apiBinding.Name, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	patchedBinding := apiBinding.DeepCopy()
+	patchedBinding.Spec.Reference.Workspace.ExportName = "other-export"
+	mergePatch, err := jsonpatch.CreateMergePatch(encodeJSON(t, apiBinding), encodeJSON(t, patchedBinding))
+	require.NoError(t, err)
+	_, err = kcpClusterClient.ApisV1alpha1().APIBindings().Patch(logicalcluster.WithCluster(ctx, consumerWorkspace), apiBinding.Name, types.MergePatchType, mergePatch, metav1.PatchOptions{})
+	require.ErrorContains(t, err, "APIExport reference must not be changed")
```

Comment:
do we rather want this as a unit test? seems a bit heavy-weight here. I saw some helper to use the CEL validator in tests.

Reply:
I'm fine to have it here

Reply:
ack, i added a unit test as well 👍